nav2_graceful_motion_controller/src/graceful_motion_controller.cpp

```diff
@@ -0,0 +1,344 @@
+{
+  path_handler_->setPlan(path);
+  goal_reached_ = false;
+}
+
+void GracefulMotionController::setSpeedLimit(
+  const double & speed_limit, const bool & percentage)
+{
+  std::lock_guard<std::mutex> param_lock(param_handler_->getMutex());
+
+  if (percentage) {
+    // Speed limit is expressed in % from maximum speed of robot
+    params_->v_linear_max = std::max(
+      params_->v_linear_max_initial * speed_limit / 100.0,
+      params_->v_linear_min);
+  } else {
+    // Speed limit is expressed in m/s
+    params_->v_linear_max = std::max(speed_limit, params_->v_linear_min);
+  }
+
+  // Update the speed limit in the control law
+  control_law_->setSpeedLimit(params_->v_linear_min, params_->v_linear_max, params_->v_angular_max);
+}
+
+geometry_msgs::msg::PoseStamped GracefulMotionController::getMotionTarget(
+  const double & motion_target_dist,
+  const nav_msgs::msg::Path & transformed_plan)
+{
+  // Find the first pose which is at a distance greater than the motion target distance
+  auto goal_pose_it = std::find_if(
+    transformed_plan.poses.begin(), transformed_plan.poses.end(), [&](const auto & ps) {
+      return std::hypot(ps.pose.position.x, ps.pose.position.y) >= motion_target_dist;
+    });
+
+  // If the pose is not far enough, take the last pose
+  if (goal_pose_it == transformed_plan.poses.end()) {
+    goal_pose_it = std::prev(transformed_plan.poses.end());
+  }
+
+  return *goal_pose_it;
+}
+
+bool GracefulMotionController::simulateTrajectory(
+  const geometry_msgs::msg::PoseStamped & robot_pose,
+  const geometry_msgs::msg::PoseStamped & motion_target, nav_msgs::msg::Path & trajectory,
+  const bool & backward)
+{
+  // Check for collision before moving
+  if (inCollision(
+      robot_pose.pose.position.x, robot_pose.pose.position.y,
+      tf2::getYaw(robot_pose.pose.orientation)))
+  {
+    return false;
+  }
+
+  // First pose
+  geometry_msgs::msg::PoseStamped next_pose;
+  next_pose.header.frame_id = costmap_ros_->getBaseFrameID();
+  next_pose.pose.orientation.w = 1.0;
+  trajectory.poses.push_back(next_pose);
+
+  double distance = std::numeric_limits<double>::max();
+  double resolution_ = costmap_ros_->getCostmap()->getResolution();
+  double dt = (params_->v_linear_max > 0.0) ? resolution_ / params_->v_linear_max : 0.0;
+
+  // Set max iter to avoid infinite loop
+  unsigned int max_iter = 2 * sqrt(
+    motion_target.pose.position.x * motion_target.pose.position.x +
+    motion_target.pose.position.y * motion_target.pose.position.y) / resolution_;
+
+  // Generate path
+  do{
+    // Apply velocities to calculate next pose
+    next_pose.pose = control_law_->calculateNextPose(
+      dt, motion_target.pose, next_pose.pose, backward);
+
+    // Add the pose to the trajectory for visualization
+    trajectory.poses.push_back(next_pose);
+
+    // Check for collision
+    geometry_msgs::msg::PoseStamped global_pose;
+    tf2::doTransform(next_pose, global_pose, costmap_transform_);
```

Comment:
Here's where I mean. What's next_pose frame? Should be odom, the local costmap frame already, right? Or is this because we're in robot-centric coordinates?